lightning-net-tokio/src/lib.rs

```diff
@@ -1,277 +1,423 @@
 use tokio::net::TcpStream;
+use tokio::{io, time};
+use tokio::sync::{mpsc, oneshot};
+use tokio::io::{AsyncReadExt, AsyncWrite, AsyncWriteExt};
 
 use lightning::ln::peer_handler;
 use lightning::ln::peer_handler::SocketDescriptor as LnSocketTrait;
 use lightning::ln::msgs::ChannelMessageHandler;
 
-use std::mem;
+use std::{task, thread};
 use std::net::SocketAddr;
 use std::sync::{Arc, Mutex};
 use std::sync::atomic::{AtomicU64, Ordering};
-use std::time::{Duration, Instant};
-use std::vec::Vec;
+use std::time::Duration;
 use std::hash::Hash;
 
 static ID_COUNTER: AtomicU64 = AtomicU64::new(0);
 
-/// A connection to a remote peer. Can be constructed either as a remote connection using
-/// Connection::setup_outbound o
-pub struct Connection {
-	writer: Option<mpsc::Sender<bytes::Bytes>>,
+/// Connection contains all our internal state for a connection - we hold a reference to the
+/// Connection object (in an Arc<Mutex<>>) in each SocketDescriptor we create as well as in the
+/// read future (which is returned by schedule_read).
+struct Connection {
+	writer: Option<io::WriteHalf<TcpStream>>,
 	event_notify: mpsc::Sender<()>,
-	pending_read: Vec<u8>,
-	read_blocker: Option<futures::sync::oneshot::Sender<Result<(), ()>>>,
+	// Because our PeerManager is templated by user-provided types, and we can't (as far as I can
+	// tell) have a const RawWakerVTable built out of templated functions, we need some indirection
+	// between being woken up with write-ready and calling PeerManager::write_event. This provides
+	// that indirection, with a Sender which gets handed to the PeerManager Arc on the
+	// schedule_read stack.
+	//
+	// An alternative (likely more effecient) approach would involve creating a RawWakerVTable at
+	// runtime with functions templated by the Arc<PeerManager> type, calling write_event directly
+	// from tokio's write wake, however doing so would require more unsafe voodo than I really feel
+	// like writing.
+	write_avail: mpsc::Sender<()>,
+	// When we are told by rust-lightning to pause read (because we have writes backing up), we do
+	// so by setting read_paused. If the read thread thereafter reads some data, it will place a
+	// Sender here and then block on it.
+	read_blocker: Option<oneshot::Sender<()>>,
+	// When we are told by rust-lightning to disconnect, we can't return to rust-lightning until we
+	// are sure we won't call any more read/write PeerManager functions with the same connection.
+	// This is set to true if we're in such a condition (with disconnect checked before with the
+	// top-level mutex held) and false when we can return.
+	disconnect_block: bool,
 	read_paused: bool,
-	need_disconnect: bool,
+	// If we get disconnected via SocketDescriptor::disconnect_socket(), we don't call
+	// disconnect_event(), but if we get an Err return value out of PeerManager, in general, we do.
+	// We track here whether we'll need to call disconnect_event() after the socket closes.
+	need_disconnect_event: bool,
```

Comment:
Wish there was a way to get rid of this variable, feels a bit hacky.

I kept having to re-reference what it means, so maybe rename more explicitly to something like "call_disconnect_event_on_shutdown"?

Reply:
Sure, renamed. Honestly after all the back and forth on it and bugs and whatnot I think I agree - we should just *always* require one, though the handling inside PeerHandler will likely get a bunch more complicated, so I don't know if it makes sense to run forward and do it ASAP.

Reply:
Doesn't seem to be renamed? Not a huge deal tho. 

>we should just always require one, though the handling inside PeerHandler will likely get a bunch more complicated, so I don't know if it makes sense to run forward and do it ASAP.

Cool, might file an issue for that then!